### PR TITLE
Validate settings before reloading AD/LDAP bind password

### PR DESCRIPTION
**Pull request by slobodanadamovic**
This is a followup to https://github.com/elastic/elasticsearch/pull/104320 which adds validation during secure setting reload of a `bind_password`. The reload will now fail with an exception instead of logging a deprecation warning .

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java [hunk 1]
```diff
@@ -104,17 +105,28 @@ private SimpleBindRequest buildBindRequest(Settings settings) {
...
-                    RealmSettings.getFullSettingKey(config, LEGACY_BIND_PASSWORD)
-                );
+                if (failOnMissingBindPassword) {
+                    throw new SettingsException(
+                        "[{}] is set but no bind password is specified. Without a corresponding bind password, "
+                            + "all {} realm authentication will fail. Specify a bind password via [{}] or [{}].",
+                        RealmSettings.getFullSettingKey(config, PoolingSessionFactorySettings.BIND_DN),
+                        config.type(),
+                        RealmSettings.getFullSettingKey(config, SECURE_BIND_PASSWORD),
+                        RealmSettings.getFullSettingKey(config, LEGACY_BIND_PASSWORD)
```
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java:115]: The only thing I'm not sure is if we want to advertise using a deprecated non-secure setting here.

**Review by jakelandis — COMMENTED** [review]
LGTM

I don't have any strong preferences for the 2 comments, and fine to ship as-is.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java (hunk 1 above)
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java:115]: Thanks for the suggestions. After thinking about this a bit more, I think the correct option is **b)**, given that `LEGACY_BIND_PASSWORD` is a node scoped, non-secure setting that cannot be dynamically changed via reload API. Unit tests were suggesting that both legacy and secure settings can be reloaded, which is wrong. So, I've changed them.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java [hunk 2]
```diff
@@ -82,7 +83,7 @@ abstract class PoolingSessionFactory extends SessionFactory implements Releasabl
         }
     }
 
-    private SimpleBindRequest buildBindRequest(Settings settings) {
+    private SimpleBindRequest buildBindRequest(Settings settings, boolean failOnMissingBindPassword) {
```
[the remark this answers is not in the document]
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java:86]: 👍 I've renamed it and added javadoc with brief explanation.

**Comment by tvernum** [comment]
Are you sure this change is safe?

I suspect (but haven't tested) that it means:

- if you have a node with the deprecated config (bind_dn, no password)
- and you call `POST /_nodes/reload_secure_settings` for **_any reason_**

then it will fail and you cannot reload the settings you were trying to reload.

Which would mean (for example) that a cluster running with this deprecated but permitted config would no longer be able to reload AWS secret keys for their S3 snapshot repository.

**Comment by slobodanadamovic** [comment]
> Are you sure this change is safe?

Thanks for raising this concern. I believe throwing validation exceptions here is safe because exceptions thrown by `Security` plugin should not affect reloading settings for other plugins:

https://github.com/elastic/elasticsearch/blob/01f19ecab295e961bb262fcc01391bcf12dd81d9/server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/TransportNodesReloadSecureSettingsAction.java#L130-L136

Also, exceptions thrown by AD/LDAP realm (or any other realm) should not affect reloading setting for other reloadable security components:

https://github.com/elastic/elasticsearch/blob/01f19ecab295e961bb262fcc01391bcf12dd81d9/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java#L573-L590

https://github.com/elastic/elasticsearch/blob/01f19ecab295e961bb262fcc01391bcf12dd81d9/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java#L1949-L1976

**Pull request merged**